### ISSUE TITLE
Bad performances on custom objective function (Exponential loss) 

### ISSUE THREAD

**Issue by jf10** [dmlc/xgboost#2072]
Hi,

I have been trying to implement the exponential loss (as in adaboost or in sklearn.GradientBoostingClassifier). However the results I get using this loss function in xgboost are significantly worse than those I get using the sklearn library. 

   ```
 def exponentialLoss(y, pred):
        df = -y * np.exp(-y * pred)
        hess = np.exp(-y * pred)
        return df, hess
```

I just set up my XGBClassifier with the sklearn api in python with the following:

`clf = XGBClassifier(objective=exponentialLoss)`

I am wondering if there is some normalization to do before sending the first and second order derivatives. Or maybe the sklearn api does not handle well the custom objective function ?

**Comment by SweeRoty** [comment]
The negative labels should be -1 instead of 0. If in R, add y <- ifelse(y == 0, -1, 1) before calculating grad and hess.

**Comment by jf10** [comment]
Thank you for your answer. I removed it in the example but at the beginning of my function I have :

y_ = (2. * y - 1.)

Then I use y_ to compute my gradient and hessian.

I can't really understand why xgboost gets such bad results compared to sklearn tbh. By the way, what is the objective function behind the "reg:linear". Is is cross entropy ?

**Comment by SweeRoty** [comment]
It seems that you should convert prediction to probabilities using logistic function, and then calculate grad and hess. Check this https://github.com/dmlc/xgboost/issues/15

**Comment by jf10** [comment]
For the exponential loss you need your prediction to also be negative. If I do that I will only have predictions > 0. It does fit with the exponential loss.
